Huffman decoder rewrite / cleanup

axxel opened the pull request:
This is pretty much a rewrite of all Huffman decoding related code.

The PR size is a lot smaller than the 'big' one recently even though it is still relatively big. There was simply no meaningful way to reduce that further. It is basically just a single new implementation in HuffmanTable.h and then the removal of all the former implementations and minor adoptions to the calling code where necessary.

For details about the changes see https://github.com/darktable-org/rawspeed/commit/b213eafc6b3d4a565e52aa97b95872adebcc1879.

TL;DR: less code + faster code => better code

axxel committed:
Buffer/ByteStream: add convenience data access methods

axxel committed:
HuffmanTable: rewrite Huffman decoder related code

This is pretty much a rewrite of all Huffman decoding related code.

There have been several copies of this functionality with minor
differences (and different bugs). They are all replaced by one new
implementation in HuffmanTable.h. Code documentation of the inner
workings has been significantly increased.

Great emphasis has been put on optimizing the hot paths of the decoding
on down to the CPU instruction level. They are responsible for somehere
nearly the complete decoding time of a standard LJPEG based raw. There
is an optimized, unified lookup mechanism, replacing the old 'numbit'
and 'bigTable' funtionality.

To give an example: the number of executed instructions and executed
branches went down by about 30% - the runtime by 10%-20% for some sample
CR2.

This cleanup is not quite complete. The LJPEGDecoder and friends clases
using this code have only been changed minally to support the new code.
This will follow in a separte commit.

LebedevRI commented:
Less code is _probably_ always better :)

axxel committed:
add missing include (only detected on mingw build)

std::accumulate is found in <numeric>

axxel commented:
> Less code is probably always better :)

True. But since I am the author of the code I am A) biased and B) convinced that _this_ code is better (in a number of metrics). :)

axxel committed:
Fix (misleading) compiler warning

According to the c++ reference manual, this code was exactly right:
http://en.cppreference.com/w/cpp/container/array

Just making some compilers be quite. sigh...

LebedevRI commented on RawSpeed/LJpegDecompressor.cpp:
```diff
@@ -223,7 +223,8 @@ void LJpegDecompressor::parseDHT() {
   // in the huffmanTableStore to be later transalted to pointers.
```
translated

LebedevRI reviewed (changes requested):
I don't think this happens with git develop branch
```
=================================================================
==1276==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x611000017470 at pc 0x55f383aac955 bp 0x7f965af2d410 sp 0x7f965af2d408
READ of size 8 at 0x611000017470 thread T140
    #0 0x55f383aac954 in std::vector<int, std::allocator<int> >::operator[](unsigned long) const /usr/include/c++/6/bits/stl_vector.h:796
    #1 0x55f383aac954 in int RawSpeed::HuffmanTable::decode<RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut>, true>(RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut>&) const /home/lebedevri/rawspeed/RawSpeed/HuffmanTable.h:232
    #2 0x55f383aac954 in int RawSpeed::HuffmanTable::decodeNext<RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut> >(RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut>&) const /home/lebedevri/rawspeed/RawSpeed/HuffmanTable.h:218
    #3 0x55f383aac954 in RawSpeed::LJpegDecompressor::HuffDecode(RawSpeed::HuffmanTable*) /home/lebedevri/rawspeed/RawSpeed/LJpegDecompressor.h:159
    #4 0x55f383aac954 in RawSpeed::LJpegPlain::decodeScanLeft2Comps() /home/lebedevri/rawspeed/RawSpeed/LJpegPlain.cpp:649
    #5 0x55f383ac4e89 in RawSpeed::LJpegPlain::decodeScan() /home/lebedevri/rawspeed/RawSpeed/LJpegPlain.cpp:117
    #6 0x55f383a9b54a in RawSpeed::LJpegDecompressor::parseSOS() /home/lebedevri/rawspeed/RawSpeed/LJpegDecompressor.cpp:210
    #7 0x55f383a9df51 in RawSpeed::LJpegDecompressor::startDecoder(unsigned int, unsigned int, unsigned int, unsigned int) /home/lebedevri/rawspeed/RawSpeed/LJpegDecompressor.cpp:95
    #8 0x55f383a8749d in RawSpeed::DngDecoderSlices::decodeSlice(RawSpeed::DngDecoderThread*) /home/lebedevri/rawspeed/RawSpeed/DngDecoderSlices.cpp:362
    #9 0x55f383a883e9 in RawSpeed::DecodeThread(void*) /home/lebedevri/rawspeed/RawSpeed/DngDecoderSlices.cpp:31
    #10 0x7f965ffbb423 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x7423)
    #11 0x7f965f2249be in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe89be)

AddressSanitizer can not describe address in more detail (wild memory access suspected).
SUMMARY: AddressSanitizer: heap-buffer-overflow /usr/include/c++/6/bits/stl_vector.h:796 in std::vector<int, std::allocator<int> >::operator[](unsigned long) const
Shadow bytes around the buggy address:
  0x0c227fffae30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffae40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffae50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffae60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffae70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c227fffae80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa[fa]fa
  0x0c227fffae90: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c227fffaea0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffaeb0: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c227fffaec0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffaed0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
Thread T140 created by T0 here:
    #0 0x7f96608d4f59 in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x30f59)
    #1 0x55f383a86c06 in RawSpeed::DngDecoderSlices::startDecoding() /home/lebedevri/rawspeed/RawSpeed/DngDecoderSlices.cpp:86
    #2 0x55f383a82a87 in RawSpeed::DngDecoder::decodeRawInternal() /home/lebedevri/rawspeed/RawSpeed/DngDecoder.cpp:306
    #3 0x55f383a3a5e9 in RawSpeed::RawDecoder::decodeRaw() /home/lebedevri/rawspeed/RawSpeed/RawDecoder.cpp:690
    #4 0x55f383a248de in process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RawSpeed::CameraMetaData*, bool, bool) /home/lebedevri/rawspeed/RawSpeed/rstest.cpp:196
    #5 0x55f383a261f7 in main /home/lebedevri/rawspeed/RawSpeed/rstest.cpp:284
    #6 0x7f965f15c2b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

==1276==ABORTING
```

axxel commented:
> I don't think this happens with git develop branch

Probably not. I can not reproduce this here. Before I pushed I had an ASAN build of rstest run through all the samples (freshly synched this afternoon). Can you tell me which file is causing this?

LebedevRI commented:
```
+ IFS=
+ read -r -d '' line
+ echo './Apple/iPhone 6s Plus/IMG_0853.DNG'
./Apple/iPhone 6s Plus/IMG_0853.DNG
+ /home/lebedevri/rawspeed/build/RawSpeed/rstest './Apple/iPhone 6s Plus/IMG_0853.DNG'
RawSpeed:Unable to find camera in database: 'Apple' 'iPhone 6s Plus' 'dng'
Please consider providing samples on <https://raw.pixls.us/>, thanks!
=================================================================
==11121==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61100001deb0 at pc 0x5583c3aa1955 bp 0x7fa05dfee410 sp 0x7fa05dfee408
READ of size 8 at 0x61100001deb0 thread T1
    #0 0x5583c3aa1954 in std::vector<int, std::allocator<int> >::operator[](unsigned long) const /usr/include/c++/6/bits/stl_vector.h:796
    #1 0x5583c3aa1954 in int RawSpeed::HuffmanTable::decode<RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut>, true>(RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut>&) const /home/lebedevri/rawspeed/RawSpeed/HuffmanTable.h:232
    #2 0x5583c3aa1954 in int RawSpeed::HuffmanTable::decodeNext<RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut> >(RawSpeed::BitStream<RawSpeed::JPEGBitPumpTag, RawSpeed::BitStreamCacheRightInLeftOut>&) const /home/lebedevri/rawspeed/RawSpeed/HuffmanTable.h:218
    #3 0x5583c3aa1954 in RawSpeed::LJpegDecompressor::HuffDecode(RawSpeed::HuffmanTable*) /home/lebedevri/rawspeed/RawSpeed/LJpegDecompressor.h:159
    #4 0x5583c3aa1954 in RawSpeed::LJpegPlain::decodeScanLeft2Comps() /home/lebedevri/rawspeed/RawSpeed/LJpegPlain.cpp:649
    #5 0x5583c3ab9e89 in RawSpeed::LJpegPlain::decodeScan() /home/lebedevri/rawspeed/RawSpeed/LJpegPlain.cpp:117
    #6 0x5583c3a9054a in RawSpeed::LJpegDecompressor::parseSOS() /home/lebedevri/rawspeed/RawSpeed/LJpegDecompressor.cpp:210
    #7 0x5583c3a92f51 in RawSpeed::LJpegDecompressor::startDecoder(unsigned int, unsigned int, unsigned int, unsigned int) /home/lebedevri/rawspeed/RawSpeed/LJpegDecompressor.cpp:95
    #8 0x5583c3a7c49d in RawSpeed::DngDecoderSlices::decodeSlice(RawSpeed::DngDecoderThread*) /home/lebedevri/rawspeed/RawSpeed/DngDecoderSlices.cpp:362
    #9 0x5583c3a7d3e9 in RawSpeed::DecodeThread(void*) /home/lebedevri/rawspeed/RawSpeed/DngDecoderSlices.cpp:31
    #10 0x7fa064553423 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x7423)
    #11 0x7fa0637bc9be in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe89be)

AddressSanitizer can not describe address in more detail (wild memory access suspected).
SUMMARY: AddressSanitizer: heap-buffer-overflow /usr/include/c++/6/bits/stl_vector.h:796 in std::vector<int, std::allocator<int> >::operator[](unsigned long) const
Shadow bytes around the buggy address:
  0x0c227fffbb80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffbb90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffbba0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffbbb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffbbc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c227fffbbd0: fa fa fa fa fa fa[fa]fa fa fa fa fa fa fa fa fa
  0x0c227fffbbe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffbbf0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c227fffbc00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffbc10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c227fffbc20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
Thread T1 created by T0 here:
    #0 0x7fa064e6cf59 in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x30f59)
    #1 0x5583c3a7bc06 in RawSpeed::DngDecoderSlices::startDecoding() /home/lebedevri/rawspeed/RawSpeed/DngDecoderSlices.cpp:86
    #2 0x5583c3a77a87 in RawSpeed::DngDecoder::decodeRawInternal() /home/lebedevri/rawspeed/RawSpeed/DngDecoder.cpp:306
    #3 0x5583c3a2f5e9 in RawSpeed::RawDecoder::decodeRaw() /home/lebedevri/rawspeed/RawSpeed/RawDecoder.cpp:690
    #4 0x5583c3a198de in process(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RawSpeed::CameraMetaData*, bool, bool) /home/lebedevri/rawspeed/RawSpeed/rstest.cpp:196
    #5 0x5583c3a1b1f7 in main /home/lebedevri/rawspeed/RawSpeed/rstest.cpp:284
    #6 0x7fa0636f42b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

==11121==ABORTING
```

LebedevRI commented:
> (freshly synched this afternoon)

thankfully, now the update procedure only downloads what is needed, so it is absolutely best to sync each time before running a full pass. or, at least before creating new pr

axxel commented:
> thankfully, now the update procedure only downloads what is needed, so it is absolutely best to sync each time before running a full pass. or, at least before creating new pr

Indeed and that is just what I did :). Well, maybe there were up to 60 minutes between my sync and the final `git push` due to the time the tests rand and I needed to write up the commit message. But I can live very well with someone sneaking in one more file in those 60 minutes :).

Thanks for pointing me to that file.

axxel committed:
LJpegDecompressor: support files with multiple DHT markers

Apple seems to think it is cool to have a DNG with LJpeg compressed data
that has 2 DHT fields with one Huffman table each instead of one DHT
field with all the Huffman tables (like _everybody_ else).

Since the pointer setup to the different Huffman tables would become
even more cryptic when supporting this behavior, I decided to switch
from storage of HuffmanTable objects to storage of
unique_ptr<HuffmanTable>. This way the pointer invalidation problem goes
away.

axxel committed:
HuffmanTable: fix performace bug

Make the comparison of HuffmanTable objects work again, which broke
sometime during development before the first commit.

This re-enables the reduction to the minimum required Huffman tables,
which should help in reducing cache pollution during decoding.

axxel committed:
HuffmanTable: performance tuning

Make the lookup table smaller (11 instead of 13 bits). This results in a
few percent performance improvement. The optimal number is likely CPU
dependent.

axxel commented:
That was a rather tricky one: Apple had fun with the DHT markers. Details see commit message. Would be interesting to know what the spec has to say about that. Anyway, I am actually glad it happened because this way I found a small performance bug that sneaked in shortly before I send in this pr.

LebedevRI reviewed (approved):
Seems to still work
